twilight-gateway/src/stream.rs

```diff
@@ -51,43 +75,349 @@ pub enum StartRecommendedErrorType {
+        };
+
+        for shard in shards {
+            this.add_shard(shard);
+        }
+
+        this
+    }
+
+    /// Add a shard to the stream to produce a gateway event.
+    fn add_shard(&mut self, shard: &'a mut Shard) {
+        self.futures.borrow_mut().push(Box::pin(async {
+            let result = shard.next_event().await;
+
+            NextItemOutput { result, shard }
+        }));
+    }
+}
+
+impl<'a> Stream for ShardEventStream<'a> {
+    type Item = Result<(ShardRef<'a>, Event), ReceiveMessageError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.as_mut();
+        let poll = this.futures.borrow_mut().poll_next_unpin(cx);
+
+        match poll {
+            Poll::Ready(Some(output)) => Poll::Ready(Some(output.result.map(|message| {
+                (
+                    ShardRef {
+                        list: ShardList::Events(Rc::clone(&this.futures)),
+                        shard: Some(output.shard),
+                    },
+                    message,
+                )
+            }))),
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+/// Stream selecting the next websocket message from a group of shards.
+///
+/// # Examples
+///
+/// Create the recommended number of shards and stream over their messages:
+///
+/// ```no_run
+/// use futures::StreamExt;
+/// use std::{collections::HashMap, env, future};
+/// use twilight_gateway::{stream::{self, ShardMessageStream}, Config, Intents};
+///
+/// # #[tokio::main]
+/// # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+/// let token = env::var("DISCORD_TOKEN")?;
+///
+/// // callback to create a config for each shard, useful for when not all shards
+/// // have the same configuration, such as for per-shard presences
+/// let config_callback = |_| Config::new(token.clone(), Intents::GUILDS);
+///
+/// let mut shards = stream::start_recommended(token.clone(), config_callback)
+///     .await?
+///     .filter_map(|shard_result| async move { shard_result.ok() })
+///     .collect::<Vec<_>>()
+///     .await;
+///
+/// let mut stream = ShardMessageStream::new(shards.iter_mut());
+///
+/// loop {
+///     let (shard, message) = match stream.next().await {
+///         Some(Ok((shard, message))) => (shard, message),
+///         Some(Err(source)) => {
+///             tracing::warn!(?source, "error receiving message");
+///
+///             if source.is_fatal() {
+///                 break;
+///             }
+///
+///             continue;
+///         },
+///         None => break,
+///     };
+///
+///     println!("received message on shard {}: {message:?}", shard.id());
+/// }
+/// # Ok(()) }
+/// ```
+pub struct ShardMessageStream<'a> {
+    /// Set of futures resolving to the next message of each shard.
+    futures: Rc<RefCell<FutureList<'a, Message>>>,
+}
```

Comment:
Would it not be best if this implements Send?

Reply:
please see my reply with additional context https://github.com/twilight-rs/twilight/pull/1809#issuecomment-1200325787